lib/mongo/retryable.rb

```diff
@@ -258,23 +267,28 @@ def write_with_retry(session, write_concern, ending_transaction = false, &block)
     # delegates to legacy_write_with_retry which performs write retries using
     # legacy logic.
     #
+    # @param [ nil | Server ] server Optional server to use with the operation.
     # @param [ nil | Session ] session Optional session to use with the operation.
     # @param [ nil | Hash | WriteConcern::Base ] write_concern The write concern.
+    # @param [ Context ] context The context for the operation.
     #
-    # @yieldparam [ Server ] server The server to which the write should be sent.
+    # @yieldparam [ Connection ] connection The connection through which the
+    #   write should be sent.
     #
     # @api private
-    def nro_write_with_retry(session, write_concern, &block)
+    def nro_write_with_retry(server, session, write_concern, context:, &block)
```

Comment:
I understand that. Since MR is looking at server description, and a connection already exists for the operation being attempted, the most correct thing to do is to examine server description on the connection. If that description is unsuitable, the query needs to be rerouted, which does require the server object but the server object can be retrieved from the connection object.